denote.el

```diff
@@ -681,13 +681,13 @@ The note's ID is derived from the date and time of its creation.")
 (defconst denote-id-regexp "\\([0-9]\\{8\\}\\)\\(T[0-9]\\{6\\}\\)"
   "Regular expression to match `denote-id-format'.")
 
-(defconst denote-signature-regexp "==\\([^.]*?\\)\\(--.*\\|__.*\\|\\..*\\)*$"
+(defconst denote-signature-regexp "==\\(.*?\\)\\(--.*\\|__.*\\)*$"
   "Regular expression to match the SIGNATURE field in a file name.")
 
-(defconst denote-title-regexp "--\\([^.]*?\\)\\(--.*\\|__.*\\|\\..*\\)*$"
+(defconst denote-title-regexp "--\\(.*?\\)\\(__.*\\|==.*\\)*$"
   "Regular expression to match the TITLE field in a file name.")
 
-(defconst denote-keywords-regexp "__\\([^.]*?\\)\\(--.*\\|__.*\\|\\..*\\)*$"
+(defconst denote-keywords-regexp "__\\(.*?\\)\\(--.*\\|__.*\\|==.*\\)*$"
```

Comment:
Bug? This regex will also capture the extension:

```lisp
(let ((file "01020304T123456--foo__qux.md"))
    (string-match "__\\(.*?\\)\\(--.*\\|__.*\\|==.*\\)*$" file)
    (match-string 1 file))
```

Reply:
I relate to the desire to enable as many characters as possible\*, but I think `.` might just be too prone to pitfalls. Especially if you want to let people change the order of fields later.

Remember also that outside of notes, people might have files like `20230000T000000--foo__bar_qux.tar.gz` (is it `qux.tar` or tag `qux` and inner extension `.tar`? what if we replace `.tar` with `.xyz`?)

\* Although over the years I have come to appreciate limitations. A good set of tight limitations empowers ecosystem contributions by simplifying the problem space of interoperability, and I think the right balance of limitations and flexibility is just creating something just barely so useful that you decide to tolerate the limitations.

Reply:
Yes, this regex, as-is, captures the extension. But I also modified the places where it is used to first strip the extension and THEN apply the regexp. The regex has already been removed on the string the we apply the regexp on.

I also understand that some extensions have multiple `.` in them, like `.tar.gz`. This is not perfect and we cannot really do better then `file-name-extension`. `file-name-extension` also fails to recognize the "right" extension here. Denote would be no better.

Reply:
By the way, it is not has crucial as one may think to parse the file name with perfect accuracy. Generally, the information in the front matter has priority (when available) over what is in the file name. Same with fontification. Even if there are some edge cases that are not perfectly handled, this does not break Denote internally. Moreover, this information that we retrieve in the file name is mostly used to provide suggestions in the creation or renaming commands.

Reply:
> it is not has crucial as one may think to parse the file name with perfect accuracy

So the issue that I see is:

1. Having a file like `foo__bar.tar.gz`
2. I call my keybind to
    * remove tags: it presents candidate `bar.tar` instead of `bar`;
    * add tags: I type/select `qux`, it renames my file to `foo__bar.tar_qux.gz`.

Reply:
You're right! I have removed that part from the pull request.